components/range_cache_memory_engine/src/range_manager.rs

```diff
@@ -199,127 +318,237 @@ impl RangeManager {
+        let entry = self
+            .regions_by_range
+            .range::<[u8], (Bound<&[u8]>, Bound<&[u8]>)>((Excluded(&range.start), Unbounded))
+            .next();
+        if let Some((_key, id)) = entry {
+            let meta = &self.regions[id];
+            if meta.range.start < range.end {
+                return true;
+            }
+        }
+        false
     }
 
-    fn overlap_with_pending_range(&self, range: &CacheRange) -> bool {
-        self.pending_ranges.iter().any(|r| r.overlaps(range))
-            || self
-                .pending_ranges_loading_data
-                .iter()
-                .any(|(r, ..)| r.overlaps(range))
+    /// `check_overlap_with_region` check of there are regions overlap with
+    /// target region. If there are regions with `pending` state and whose
+    /// epoch verion is smaller than target region, the pending regions will
+    /// be removed first.
+    fn check_overlap_with_region(&mut self, region: &Region) -> Option<RegionState> {
+        let region_range = CacheRange::from_region(region);
+        let mut removed_region = vec![];
+        let mut overlapped_region_state = None;
+        self.iter_overlapped_regions(&region_range, |region_meta| {
+            if region_meta.state != RegionState::Pending
+                || region_meta.region.get_region_epoch().version
+                    >= region.get_region_epoch().version
+            {
+                tikv_util::warn!("load region overlaps with existing region"; "region" => ?region, "exist_meta" => ?region_meta);
+                overlapped_region_state = Some(region_meta.state);
+                return false;
+            }
+            // pending region with out-dated epoch, should be removed.
+            removed_region.push(region_meta.region.id);
+            true
+        });
+        for id in removed_region {
+            self.remove_region(id);
+        }
+        overlapped_region_state
     }
 
     // Acquire a snapshot of the `range` with `read_ts`. If the range is not
     // accessable, None will be returned. Otherwise, the range id will be returned.
-    pub(crate) fn range_snapshot(
+    pub(crate) fn region_snapshot(
         &mut self,
-        range: &CacheRange,
+        region_id: u64,
+        region_epoch: u64,
         read_ts: u64,
-    ) -> result::Result<u64, FailedReason> {
-        let Some(range_key) = self
-            .ranges
-            .keys()
-            .find(|&r| r.contains_range(range))
-            .cloned()
-        else {
+    ) -> result::Result<(), FailedReason> {
+        let Some(meta) = self.regions.get_mut(&region_id) else {
             return Err(FailedReason::NotCached);
         };
-        let meta = self.ranges.get_mut(&range_key).unwrap();
+
+        if meta.state != RegionState::Active {
+            return Err(FailedReason::NotCached);
+        }
+
+        if meta.region.get_region_epoch().version != region_epoch {
+            return Err(FailedReason::EpochNotMatch);
```

Comment:
What will happen if:
1) region_epoch is older than current meta.region.epoch?
2) region_epoch is newer than current meta.region.epoch? Does it will trigger the update region operation?

Reply:
Here, cached region epoch can only be higher than raft's region epoch if not equal. Because cached region's epoch  is updated when write batch is apply, but raft's local reader's epoch is updated after ApplyRes returned, so during this small interval, cached epoch can be higher than raft's. So during this interval, foreground request will fallback to read from rocksdb due to epoch not match.

Reply:
We can add a assert here and add some comments, so we can change this != to >